walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.ListIterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Initiatize return data structure and set duration parameter
+    Collection<TimeRange> availTimes = new HashSet<>();
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+    final int length = (int) request.getDuration();
+    Collection<String>  requestAttendees = request.getAttendees();
+    
+    // Create a List of relevant events that will restrict available time ranges
+    bookTimes = findTimeConflicts(requestAttendees, events);
+
+    // Check that there are booked meetings before sorting it and running the available function
+    if (bookTimes.isEmpty()) {
+            availTimes.add(bookTimes.get(0).WHOLE_DAY);
+    }
+
+    else {
+        Collections.sort(bookTimes, bookTimes.get(0).ORDER_BY_START);
+        availTimes = availableTimes(bookTimes, length);
+    }
+
+    return availTimes;
+  }
+
+  public ArrayList<TimeRange> findTimeConflicts(Collection<String> requestAttendees, Collection<Event> events){
+    // Create a List of relevant events that will restrict available time ranges
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+
+    for (Iterator<Event> eventIterator = events.iterator(); eventIterator.hasNext();){
+        Event event = eventIterator.next();
+      
+        for (Iterator<String> attendIterator = requestAttendees.iterator(); attendIterator.hasNext();){
+            String attendee = attendIterator.next();
+           
+            if(event.getAttendees().contains(attendee) && !bookTimes.contains(event.getWhen())){
+                bookTimes.add(event.getWhen());
+            }
+        }
+    }
+    return bookTimes;
+  }
+
+  public Collection<TimeRange> availableTimes(ArrayList<TimeRange> bookTimes, int length){
+    Collection<TimeRange> availTimes = new HashSet<>();
+    if(length > bookTimes.get(0).END_OF_DAY) {
+        return availTimes;
+    }
+// My problems are in the for loop.I believe that my conditionals are inefficient and I'm still working on it.
+    else {
+    
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();) {
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(bookIterator.hasPrevious()) {
+                TimeRange prevBookedTime = bookIterator.previous();
```

Comment:
Ah ha, I think I found your bug.

When you call "previous" it actually moves the iterator backwards! So on Line 77, you move the iterator forwards, and then Line 80 moves it backwards again, so you never actually manage to advance the iterator.

If you want to actually store the previous TimeRange, then this might work:

```
TimeRange prevBookedTime = null;
for (ListIIterator<TimeRange> bookIterator = /* like before */) {
  TimeRange bookedTime = bookIterator.next();
  if (prevBookedTime != null) {  // instead of bookIterator.hasPrevious()
     if (!bookedTime.overlaps(prevBookedTime) /* ... like before ... (/ ) {
     }
  }
 
  // Store bookedTime in prevBookedTime for the next iteration through the loop.
  prevBookedTime = bookedTime;
}
```

https://docs.oracle.com/javase/7/docs/api/java/util/ListIterator.html#previous()

Reply:
Wow thank you! When I kept looking into this error I couldn't find an example when .previous() and .next() were used in the same iterator instance. But this makes sense, pointers are a tricky thing I'm still learning about.